File: apple/Sources/ShareKit/PendingFileOffer.swift
import Foundation

/// 一个等待用户决定的入站文件请求。由 ShareEngine 在收到 FILE_OFFER 时投递。
public struct PendingFileOffer: Identifiable, Sendable, Equatable {
    public let id: UUID                      // = transfer_id
    public let peer: Device
    public let fileName: String
    public let fileSize: UInt64
    public let sha256: String
    public let receivedAt: Date

    public init(
        id: UUID,
        peer: Device,
        fileName: String,
        fileSize: UInt64,
        sha256: String,
        receivedAt: Date = Date()
    ) {
        self.id = id
        self.peer = peer
        self.fileName = fileName
        self.fileSize = fileSize
        self.sha256 = sha256
        self.receivedAt = receivedAt
    }

    public var formattedSize: String {
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = [.useKB, .useMB, .useGB]
        formatter.countStyle = .file
        return formatter.string(fromByteCount: Int64(fileSize))
    }
}

File: android/app/src/main/java/drop/mesh/data/PendingFileOffer.kt
package drop.mesh.data

import java.text.NumberFormat
import java.util.UUID

/** 等待用户决定的入站文件请求。 */
data class PendingFileOffer(
    val id: UUID,                // = transfer_id
    val peer: Device,
    val fileName: String,
    val fileSize: Long,
    val sha256: String,
    val receivedAt: Long = System.currentTimeMillis(),
) {
    val formattedSize: String
        get() = formatBytes(fileSize)
}

fun formatBytes(n: Long): String {
    if (n < 1024) return "$n B"
    val kb = n / 1024.0
    if (kb < 1024) return "${NumberFormat.getInstance().apply { maximumFractionDigits = 1 }.format(kb)} KB"
    val mb = kb / 1024.0
    if (mb < 1024) return "${NumberFormat.getInstance().apply { maximumFractionDigits = 1 }.format(mb)} MB"
    val gb = mb / 1024.0
    return "${NumberFormat.getInstance().apply { maximumFractionDigits = 2 }.format(gb)} GB"
}
